### PR TITLE
allow for basic configuration of HttpClient

### PR DESCRIPTION
**Pull request by mattnworb**
Allow the underlying OkHttpClient instance to be configured with a few
basic properties from the Config, if specified. The configurable
properties are:

- connect timeout - `http.client.connectTimeout`
- read timeout - `http.client.readTimeout`
- write timeout - `http.client.writeTimeout`
- connection pool: max idle connections - `http.client.maxIdleConnections`
- connection pool: keep alive duration - `http.client.keepAliveDuration`
- max async requests - `http.client.async.maxRequests`
- max async requests per host - `http.client.async.maxRequestsPerHost`

The default value from OkHttpClient is used for any unspecified
properties (which is the behavior previous to this commit).

Also renamed the factory method `HttpClient.create()` to
`HttpClient.createDefault()` to make it clear that the instance being
returned is using all of the default configuration properties.

### CODE REVIEW

**Comment by mattnworb** [comment]
I am not that familiar with Guice so I would appreciate any review of how I am using it here. 

For instance, it feels weird to me that I had to create a HttpClientProvider when all it does is wire the OkHttpClient instance into the HttpClient constructor. Is there a way to avoid declaring a class for just that? I originally had all of the OkHttpClient construction in the HttpClientProvider class, but I wanted to be able to add a test that the Config keys were properly parsed into custom values of things like `OkHttpClient.getConnectTimeout()` but couldn't see an easier way to do that.

Also, is it problematic to have something like `bind(OkHttpClient.class).to(OkHttpClientProvider.class)` when the former is a 3rd party class that someone could conceivably already be binding in their own service?

**Comment by mattnworb** [comment]
btw, the license-maven-plugin seems to enforce that the license header has trailing spaces in lines that only contain `*`, which some editors might automatically remove 😕

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecoratorProvider.java [hunk 1]
```diff
@@ -21,12 +21,21 @@
...
 import javax.inject.Provider;
 
 class HttpClientDecoratorProvider implements Provider<ClientDecorator> {
 
+  private final HttpClient httpClient;
+
+  @Inject
+  HttpClientDecoratorProvider(final HttpClient httpClient) {
+    this.httpClient = httpClient;
+  }
```
Comment (mattnworb) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecoratorProvider.java]: this also feels weird for the same reason as HttpClientProvider

**Comment by pettermahlen** [comment]
I'm also not a Guice expert, but LGTM. @rouzwawi ?

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecorator.java [hunk 2]
```diff
@@ -24,17 +24,14 @@
...
 
-  private static final HttpClientDecorator INSTANCE = new HttpClientDecorator();
+  private final HttpClient httpClient;
 
-  HttpClientDecorator() {
-  }
-
-  public static HttpClientDecorator create() {
-    return INSTANCE;
+  HttpClientDecorator(HttpClient httpClient) {
```
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecorator.java]: `@Inject` on this and you can get rid of the providers, I think.

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientProvider.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  private final OkHttpClient okHttpClient;
+
+  @Inject
+  HttpClientProvider(OkHttpClient okHttpClient) {
+    this.okHttpClient = okHttpClient;
+  }
+
+  @Override
+  public HttpClient get() {
+    return new HttpClient(okHttpClient);
```
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientProvider.java]: This constructor already has an `@Inject` annotation, so it should suffice to just bind the class in the module, as `bind(HttpClient.class)`.

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientModule.java [hunk 4]
```diff
@@ -38,6 +39,9 @@ public static ApolloModule create() {
   protected void configure() {
     Multibinder.newSetBinder(binder(), ClientDecorator.class)
         .addBinding().toProvider(HttpClientDecoratorProvider.class);
+
+    bind(HttpClient.class).toProvider(HttpClientProvider.class);
+    bind(OkHttpClient.class).toProvider(OkHttpClientProvider.class);
```
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientModule.java]: Binding these without the `.to...` part should make Guice find injectable constructors.

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java [hunk 5]
```diff
@@ -38,11 +40,18 @@
...
     this.client = client;
   }
 
+  @Deprecated
+  /** @deprecated use {@link #createDefault()} instead. */
   public static HttpClient create() {
+    return createDefault();
+  }
+
+  public static HttpClient createDefault() {
```
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java]: Why did this have to change name?

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java [hunk 6]
```diff
@@ -20,6 +20,8 @@
 package com.spotify.apollo.http.client;
 
 import com.spotify.apollo.environment.IncomingRequestAwareClient;
+
+import com.google.inject.Inject;
```
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java]: prefer using `javax.inject.Inject`

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientProvider.java (hunk 3 above)
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientProvider.java]: As done here: https://github.com/spotify/apollo/blob/master/modules/jetty-http-server/src/main/java/com/spotify/apollo/http/server/HttpServerModule.java#L51

**Comment by pettermahlen** [comment]
Another thing: it would be great to add documentation of the configuration settings in a README for the client. There's no such readme at the moment.

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java (hunk 5 above)
Comment (mattnworb) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java]: It didn't have to, but I wanted to make it clear that it was creating a HttpClient that was unconfigured / with all of the default configuration values

Comment (mattnworb) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java]: and now that I say that I am not sure why I didn't just call it `createUnconfigured()` 😄

**Comment by mattnworb** [comment]
wow coveralls is precise

> ❌  Coverage decreased (-0.05%) to 67.542%

**Comment by mattnworb** [comment]
@rouzwawi @pettermahlen thanks for the review, I added new commits which simplified things and removed the need for one of the Providers, and also added a README for the module.

**Comment by pettermahlen** [comment]
👍 from me

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecoratorProvider.java (hunk 1 above)
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecoratorProvider.java]: Can't this class be dropped all together now that HttpClientDecorator has an injectable constructor?

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java (hunk 5 above)
Comment (rouzwawi) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClient.java]: Ah, makes sense, and `createUnconfigured()` makes more sense :)

File: modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecoratorProvider.java (hunk 1 above)
Comment (mattnworb) [inline modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/HttpClientDecoratorProvider.java]: Looks like you are correct - see the newest commit 9911529

**Pull request merged**